File: aurigraph-av10-7/basicnode/src/main/java/io/aurigraph/basicnode/compliance/PerformanceMonitor.java
package io.aurigraph.basicnode.compliance;

import io.micrometer.core.instrument.Counter;
import io.micrometer.core.instrument.Gauge;
import io.micrometer.core.instrument.MeterRegistry;
import io.micrometer.core.instrument.Timer;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import org.jboss.logging.Logger;

import java.lang.management.ManagementFactory;
import java.lang.management.MemoryMXBean;
import java.lang.management.OperatingSystemMXBean;
import java.time.Duration;
import java.time.Instant;
import java.util.concurrent.atomic.AtomicLong;

/**
 * AV10-17 Performance Monitor
 * Monitors node performance metrics for compliance validation
 */
@ApplicationScoped
public class PerformanceMonitor {
    
    private static final Logger LOG = Logger.getLogger(PerformanceMonitor.class);
    
    @Inject
    MeterRegistry meterRegistry;
    
    private final MemoryMXBean memoryBean = ManagementFactory.getMemoryMXBean();
    private final OperatingSystemMXBean osBean = ManagementFactory.getOperatingSystemMXBean();
    
    private final AtomicLong transactionCounter = new AtomicLong(0);
    private final AtomicLong lastTpsCalculation = new AtomicLong(System.currentTimeMillis());
    private final AtomicLong lastTransactionCount = new AtomicLong(0);
    
    private Counter transactionTotal;
    private Timer transactionTimer;
    private Gauge memoryUsageGauge;
    private Gauge cpuUsageGauge;
    private Gauge uptimeGauge;
    
    private final Instant startTime = Instant.now();
    
    public void initializeMetrics() {
        LOG.info("Initializing AV10-17 performance metrics...");
        
        // Transaction metrics
        transactionTotal = Counter.builder("aurigraph.transactions.total")
            .description("Total transactions processed")
            .tag("compliance", "AV10-17")
            .register(meterRegistry);
            
        transactionTimer = Timer.builder("aurigraph.transactions.duration")
            .description("Transaction processing time")
            .tag("compliance", "AV10-17")
            .register(meterRegistry);
        
        // Memory usage gauge
        memoryUsageGauge = Gauge.builder("aurigraph.memory.usage.mb")
            .description("Current memory usage in MB")
            .tag("compliance", "AV10-17")
            .register(meterRegistry, this, PerformanceMonitor::getCurrentMemoryUsageMB);
        
        // CPU usage gauge
        cpuUsageGauge = Gauge.builder("aurigraph.cpu.usage.percent")
            .description("Current CPU usage percentage")
            .tag("compliance", "AV10-17")
            .register(meterRegistry, this, PerformanceMonitor::getCurrentCpuUsage);
        
        // Uptime gauge
        uptimeGauge = Gauge.builder("aurigraph.uptime.seconds")
            .description("Node uptime in seconds")
            .tag("compliance", "AV10-17")
            .register(meterRegistry, this, PerformanceMonitor::getUptimeSeconds);
        
        LOG.info("AV10-17 performance metrics initialized successfully");
    }
    
    public long getCurrentMemoryUsageMB() {
        long usedMemory = memoryBean.getHeapMemoryUsage().getUsed() + 
                         memoryBean.getNonHeapMemoryUsage().getUsed();
        return usedMemory / (1024 * 1024); // Convert to MB
    }
    
    public double getCurrentCpuUsage() {
        if (osBean instanceof com.sun.management.OperatingSystemMXBean sunOsBean) {
            return sunOsBean.getProcessCpuLoad() * 100.0;
        }
        return 0.0; // Fallback for non-Sun JVMs
    }
    
    public long getCurrentTPS() {
        long currentTime = System.currentTimeMillis();
        long currentCount = transactionCounter.get();
        
        long lastTime = lastTpsCalculation.get();
        long lastCount = lastTransactionCount.get();
        
        if (currentTime - lastTime < 1000) {
            return 0; // Too early to calculate meaningful TPS
        }
        
        long timeDelta = currentTime - lastTime;
        long countDelta = currentCount - lastCount;
        
        long tps = (countDelta * 1000) / timeDelta;
        
        // Update for next calculation
        lastTpsCalculation.set(currentTime);
        lastTransactionCount.set(currentCount);
        
        return tps;
    }
    
    public double getUptimeSeconds() {
        return Duration.between(startTime, Instant.now()).getSeconds();
    }
    
    public void recordTransaction(Duration processingTime) {
        transactionCounter.incrementAndGet();
        transactionTotal.increment();
        transactionTimer.record(processingTime);
    }
    
    public AV1017PerformanceReport generatePerformanceReport() {
        AV1017PerformanceReport report = new AV1017PerformanceReport();
        
        report.nodeId = System.getProperty("aurigraph.node.id", "unknown");
        report.reportTime = Instant.now();
        report.startTime = startTime;
        report.uptime = Duration.between(startTime, Instant.now());
        
        // Current metrics
        report.memoryUsageMB = getCurrentMemoryUsageMB();
        report.cpuUsagePercent = getCurrentCpuUsage();
        report.currentTPS = getCurrentTPS();
        report.totalTransactions = transactionCounter.get();
        
        // Compliance validation
        report.memoryCompliant = report.memoryUsageMB <= maxMemoryMB;
        report.performanceCompliant = report.currentTPS >= targetTPS || report.totalTransactions < 100;
        report.uptimeCompliant = calculateUptimePercent() >= uptimeTargetPercent;
        
        // Overall compliance
        report.av1017Compliant = report.memoryCompliant && 
                                report.performanceCompliant && 
                                report.uptimeCompliant;
        
        // JVM information
        report.javaVersion = System.getProperty("java.version");
        report.jvmName = System.getProperty("java.vm.name");
        report.quarkusVersion = getQuarkusVersion();
        
        return report;
    }
    
    private double calculateUptimePercent() {
        long totalTime = Duration.between(startTime, Instant.now()).toMillis();
        if (totalTime == 0) return 100.0;
        
        long actualUptime = totalTime; // Simplified - could track actual downtime
        return (double) actualUptime / totalTime * 100.0;
    }
    
    public boolean validateAV1017Compliance() {
        AV1017PerformanceReport report = generatePerformanceReport();
        
        if (!report.av1017Compliant) {
            LOG.errorf("AV10-17 COMPLIANCE VIOLATION:");
            LOG.errorf("- Memory: %dMB (limit: %dMB) - %s", 
                report.memoryUsageMB, maxMemoryMB, 
                report.memoryCompliant ? "PASS" : "FAIL");
            LOG.errorf("- Performance: %d TPS (target: %d+) - %s", 
                report.currentTPS, targetTPS, 
                report.performanceCompliant ? "PASS" : "FAIL");
            LOG.errorf("- Uptime: %.4f%% (target: %.2f%%) - %s", 
                calculateUptimePercent(), uptimeTargetPercent,
                report.uptimeCompliant ? "PASS" : "FAIL");
            return false;
        }
        
        LOG.infof("AV10-17 COMPLIANCE: ALL REQUIREMENTS SATISFIED");
        return true;
    }
    
    private String getQuarkusVersion() {
        try {
            Package pkg = Package.getPackage("io.quarkus.runtime");
            return pkg != null ? pkg.getImplementationVersion() : "unknown";
        } catch (Exception e) {
            return "unknown";
        }
    }
    
    private void recordComplianceMetric(String name, ComplianceStatus status, String description) {
        ComplianceMetric metric = new ComplianceMetric();
        metric.name = name;
        metric.status = status;
        metric.description = description;
        metric.timestamp = Instant.now();
        metrics.put(name, metric);
    }
    
    public static class AV1017PerformanceReport {
        public String nodeId;
        public Instant reportTime;
        public Instant startTime;
        public Duration uptime;
        
        // Current metrics
        public long memoryUsageMB;
        public double cpuUsagePercent;
        public long currentTPS;
        public long totalTransactions;
        
        // Compliance status
        public boolean memoryCompliant;
        public boolean performanceCompliant;
        public boolean uptimeCompliant;
        public boolean av1017Compliant;
        
        // System information
        public String javaVersion;
        public String jvmName;
        public String quarkusVersion;
    }
    
    public static class ComplianceMetric {
        public String name;
        public AV1017ComplianceManager.ComplianceStatus status;
        public String description;
        public Instant timestamp;
    }
}

File: aurigraph-av10-7/src/monitoring/PerformanceMonitor.ts
/**
 * AV11-16 Performance Monitor - TypeScript Implementation
 * Monitors node performance metrics for AV11-17 compliance validation
 * Based on Java implementation: basicnode/src/main/java/io/aurigraph/basicnode/compliance/PerformanceMonitor.java
 */

import { EventEmitter } from 'events';
import * as os from 'os';
import * as process from 'process';

export enum ComplianceStatus {
    COMPLIANT = 'COMPLIANT',
    NON_COMPLIANT = 'NON_COMPLIANT',
    WARNING = 'WARNING',
    UNKNOWN = 'UNKNOWN'
}

export interface ComplianceMetric {
    name: string;
    status: ComplianceStatus;
    description: string;
    timestamp: Date;
    value?: number;
    threshold?: number;
}

export interface AV1117PerformanceReport {
    nodeId: string;
    reportTime: Date;
    startTime: Date;
    uptime: number; // in seconds
    
    // Current metrics
    memoryUsageMB: number;
    cpuUsagePercent: number;
    currentTPS: number;
    totalTransactions: number;
    
    // Compliance status
    memoryCompliant: boolean;
    performanceCompliant: boolean;
    uptimeCompliant: boolean;
    av1017Compliant: boolean;
    
    // System information
    nodeVersion: string;
    platform: string;
    architecture: string;
}

export interface PerformanceThresholds {
    maxMemoryMB: number;
    targetTPS: number;
    uptimeTargetPercent: number;
    cpuWarningThreshold: number;
    cpuCriticalThreshold: number;
}

export class PerformanceMonitor extends EventEmitter {
    private transactionCounter: number = 0;
    private lastTpsCalculation: number = Date.now();
    private lastTransactionCount: number = 0;
    private readonly startTime: Date = new Date();
    private metrics: Map<string, ComplianceMetric> = new Map();
    private monitoringInterval?: NodeJS.Timeout;
    
    // AV11-17 Compliance Thresholds
    private readonly thresholds: PerformanceThresholds = {
        maxMemoryMB: 4096,        // 4GB memory limit
        targetTPS: 1000000,       // 1M+ TPS target
        uptimeTargetPercent: 99.9, // 99.9% uptime requirement
        cpuWarningThreshold: 80,   // 80% CPU warning
        cpuCriticalThreshold: 95   // 95% CPU critical
    };
    
    private nodeId: string;
    
    constructor(nodeId: string = process.env.NODE_ID || 'unknown', thresholds?: Partial<PerformanceThresholds>) {
        super();
        this.nodeId = nodeId;
        
        if (thresholds) {
            this.thresholds = { ...this.thresholds, ...thresholds };
        }
        
        this.initializeMetrics();
    }
    
    private initializeMetrics(): void {
        console.log(`[${this.nodeId}] Initializing AV11-17 performance metrics...`);
        
        // Start monitoring interval (every 5 seconds)
        this.monitoringInterval = setInterval(() => {
            this.collectMetrics();
        }, 5000);
        
        console.log(`[${this.nodeId}] AV11-17 performance metrics initialized successfully`);
        this.emit('metrics-initialized');
    }
    
    private collectMetrics(): void {
        const report = this.generatePerformanceReport();
        
        // Update individual compliance metrics
        this.recordComplianceMetric(
            'memory_usage',
            report.memoryCompliant ? ComplianceStatus.COMPLIANT : ComplianceStatus.NON_COMPLIANT,
            `Memory usage: ${report.memoryUsageMB}MB (limit: ${this.thresholds.maxMemoryMB}MB)`,
            report.memoryUsageMB,
            this.thresholds.maxMemoryMB
        );
        
        this.recordComplianceMetric(
            'performance_tps',
            report.performanceCompliant ? ComplianceStatus.COMPLIANT : ComplianceStatus.NON_COMPLIANT,
            `TPS: ${report.currentTPS} (target: ${this.thresholds.targetTPS}+)`,
            report.currentTPS,
            this.thresholds.targetTPS
        );
        
        this.recordComplianceMetric(
            'uptime_percent',
            report.uptimeCompliant ? ComplianceStatus.COMPLIANT : ComplianceStatus.NON_COMPLIANT,
            `Uptime: ${this.calculateUptimePercent().toFixed(4)}% (target: ${this.thresholds.uptimeTargetPercent}%)`,
            this.calculateUptimePercent(),
            this.thresholds.uptimeTargetPercent
        );
        
        // CPU monitoring (not part of AV11-17 but useful)
        const cpuStatus = report.cpuUsagePercent >= this.thresholds.cpuCriticalThreshold 
            ? ComplianceStatus.NON_COMPLIANT
            : report.cpuUsagePercent >= this.thresholds.cpuWarningThreshold
            ? ComplianceStatus.WARNING
            : ComplianceStatus.COMPLIANT;
            
        this.recordComplianceMetric(
            'cpu_usage',
            cpuStatus,
            `CPU usage: ${report.cpuUsagePercent.toFixed(2)}%`,
            report.cpuUsagePercent,
            this.thresholds.cpuWarningThreshold
        );
        
        this.emit('metrics-updated', report);
    }
    
    public getCurrentMemoryUsageMB(): number {
        const memUsage = process.memoryUsage();
        const totalMemory = memUsage.heapUsed + memUsage.external + memUsage.arrayBuffers;
        return Math.round(totalMemory / (1024 * 1024));
    }
    
    public getCurrentCpuUsage(): number {
        const cpus = os.cpus();
        let totalIdle = 0;
        let totalTick = 0;
        
        cpus.forEach((cpu) => {
            for (const type in cpu.times) {
                totalTick += cpu.times[type as keyof typeof cpu.times];
            }
            totalIdle += cpu.times.idle;
        });
        
        return 100 - Math.round((100 * totalIdle) / totalTick);
    }
    
    public getCurrentTPS(): number {
        const currentTime = Date.now();
        const currentCount = this.transactionCounter;
        
        const timeDelta = currentTime - this.lastTpsCalculation;
        const countDelta = currentCount - this.lastTransactionCount;
        
        if (timeDelta < 1000) {
            return 0; // Too early to calculate meaningful TPS
        }
        
        const tps = Math.round((countDelta * 1000) / timeDelta);
        
        // Update for next calculation
        this.lastTpsCalculation = currentTime;
        this.lastTransactionCount = currentCount;
        
        return tps;
    }
    
    public getUptimeSeconds(): number {
        return Math.floor((Date.now() - this.startTime.getTime()) / 1000);
    }
    
    private calculateUptimePercent(): number {
        const totalTime = Date.now() - this.startTime.getTime();
        if (totalTime === 0) return 100.0;
        
        // Simplified - could track actual downtime
        const actualUptime = totalTime;
        return (actualUptime / totalTime) * 100.0;
    }
    
    public recordTransaction(processingTimeMs?: number): void {
        this.transactionCounter++;
        
        if (processingTimeMs) {
            this.emit('transaction-recorded', {
                transactionId: this.transactionCounter,
                processingTime: processingTimeMs,
                timestamp: new Date()
            });
        }
    }
    
    public generatePerformanceReport(): AV1117PerformanceReport {
        const memoryUsage = this.getCurrentMemoryUsageMB();
        const cpuUsage = this.getCurrentCpuUsage();
        const currentTPS = this.getCurrentTPS();
        const uptimePercent = this.calculateUptimePercent();
        
        const report: AV1117PerformanceReport = {
            nodeId: this.nodeId,
            reportTime: new Date(),
            startTime: this.startTime,
            uptime: this.getUptimeSeconds(),
            
            // Current metrics
            memoryUsageMB: memoryUsage,
            cpuUsagePercent: cpuUsage,
            currentTPS: currentTPS,
            totalTransactions: this.transactionCounter,
            
            // Compliance validation
            memoryCompliant: memoryUsage <= this.thresholds.maxMemoryMB,
            performanceCompliant: currentTPS >= this.thresholds.targetTPS || this.transactionCounter < 100,
            uptimeCompliant: uptimePercent >= this.thresholds.uptimeTargetPercent,
            av1017Compliant: false, // Will be set below
            
            // System information
            nodeVersion: process.version,
            platform: process.platform,
            architecture: process.arch
        };
        
        // Overall compliance
        report.av1017Compliant = report.memoryCompliant && 
                                 report.performanceCompliant && 
                                 report.uptimeCompliant;
        
        return report;
    }
    
    public validateAV1117Compliance(): boolean {
        const report = this.generatePerformanceReport();
        
        if (!report.av1017Compliant) {
            console.error(`[${this.nodeId}] AV11-17 COMPLIANCE VIOLATION:`);
            console.error(`- Memory: ${report.memoryUsageMB}MB (limit: ${this.thresholds.maxMemoryMB}MB) - ${report.memoryCompliant ? 'PASS' : 'FAIL'}`);
            console.error(`- Performance: ${report.currentTPS} TPS (target: ${this.thresholds.targetTPS}+) - ${report.performanceCompliant ? 'PASS' : 'FAIL'}`);
            console.error(`- Uptime: ${this.calculateUptimePercent().toFixed(4)}% (target: ${this.thresholds.uptimeTargetPercent}%) - ${report.uptimeCompliant ? 'PASS' : 'FAIL'}`);
            
            this.emit('compliance-violation', report);
            return false;
        }
        
        console.log(`[${this.nodeId}] AV11-17 COMPLIANCE: ALL REQUIREMENTS SATISFIED`);
        this.emit('compliance-validated', report);
        return true;
    }
    
    private recordComplianceMetric(name: string, status: ComplianceStatus, description: string, value?: number, threshold?: number): void {
        const metric: ComplianceMetric = {
            name,
            status,
            description,
            timestamp: new Date(),
            value,
            threshold
        };
        
        this.metrics.set(name, metric);
        this.emit('compliance-metric', metric);
    }
    
    public getComplianceMetrics(): ComplianceMetric[] {
        return Array.from(this.metrics.values());
    }
    
    public getComplianceMetric(name: string): ComplianceMetric | undefined {
        return this.metrics.get(name);
    }
    
    public updateThresholds(newThresholds: Partial<PerformanceThresholds>): void {
        Object.assign(this.thresholds, newThresholds);
        console.log(`[${this.nodeId}] Updated performance thresholds:`, this.thresholds);
        this.emit('thresholds-updated', this.thresholds);
    }
    
    public getThresholds(): PerformanceThresholds {
        return { ...this.thresholds };
    }
    
    public startMonitoring(): void {
        if (!this.monitoringInterval) {
            this.initializeMetrics();
            console.log(`[${this.nodeId}] Performance monitoring started`);
        }
    }
    
    public stopMonitoring(): void {
        if (this.monitoringInterval) {
            clearInterval(this.monitoringInterval);
            this.monitoringInterval = undefined;
            console.log(`[${this.nodeId}] Performance monitoring stopped`);
            this.emit('monitoring-stopped');
        }
    }
    
    public destroy(): void {
        this.stopMonitoring();
        this.removeAllListeners();
    }
}